### ISSUE TITLE
Selection mechanism: "program slicing"

### ISSUE THREAD

**Issue by edsko** [well-typed/hs-bindgen#13]
In addition to the [`bindgen`-inspired selection mechanisms](https://github.com/well-typed/hs-bindgen/issues/12), we should also have a mode where the user can choose a set of _functions_ to be interested in, and then we transitively compute which data types with `Storable` instances should be generated in order to be able to use those functions.

**Comment by dschrempf** [comment]
I think we can extend this ticket to cover all declarations, and not only functions. When a user selects a declaration, transitive dependencies should also be selected. For example,

```c
#include <stdint.h>

struct foo {
    uint64_t sixty_four;
    uint32_t thirty_two;
};
```

when selecting `foo`, we should also select `uint64_t`, and `uint32_t`.

**Comment by edsko** [comment]
When we slice, we should take external bindings into account. "Give me everything that I need for this header, but here are some external bindings you can use, no need to generate bindings for those".

**Comment by dschrempf** [comment]
Some information I have gathered:

#### External binding specifications

I think the "Give me everything that I need for this header, but here are some external bindings you can use, no need to generate bindings for those" is the default mode of operation. Even though we state in the declaration of `TranslationUnit` that

``` haskell
-- This includes all declarations from all headers that we have processed,
-- except
--
-- * declarations that were filtered out by a selection predicate
-- * declarations for which we have existing external bindings
-- * declarations that were filtered out by a binding specification
unitDecls :: [Decl p]
```

we do not even have access to configuration/external binding specifications in the `Parse` pass, and so collect all selected declarations. Declarations with external binding specifications or omissions stated in the configuration binding specification are only applied several passes later (`resolveBindingSpec`).

That is, if we also select transitive dependencies, we should be fine.

However, there is also a (possibly ?) contradictory comment in `Decl.hs`:

> We need to keep track of skipped declarations so that they can be
> given external bindings.

The declarations collected in `NonSelectedDecls` are then used in `ResolveBindingSpec`.

I think I need more context to understand what's going on. If we do not select a declaration, why do we need to even know about it in later stages?

#### Selecting transitive dependencies

In the `Parse` pass in `foldDecl` we use `evalPredicate` to decide if a declaration is selected, or not. We could extend this check in that it walks the dependency graph up one level, and check if the parent declaration was selected. This would, of course, only work if we process the parent declarations before the child declarations. Is this the case?

Another way would be that we store the parent declaration, if there is one, in the environment of the `Fold`. If there is a parent declaration, it must be selected, and so also the child.

Thinking about this a bit more: We can most likely not assume that we process declarations from parents to children. So maybe we need to use the `use-decl` graph, but that is only available after the `Sort` pass.

Do you have other ideas about how we could decide if a declaration should be selected because another declaration depends on it?

**Comment by TravisCardwell** [comment]
**External binding specifications**

For reference, here is an overview of the current phases and `Frontend` passes:

1. `Frontend` Phase
    1. `Parse`
    2. `Sort`
    3. `HandleMacros`
    4. `NameAnon`
    5. `ResolveBindingSpec`
    6. `HandleTypedefs`
    7. `MangleNames`
2. `Hs` Phase
3. `SHs` Phase
4. `Backend` Phase
    * Preprocessor
    * Template Haskell

It is worth noting that the order of passes is strictly determined by constraints.  `Sort` must come before the following passes because we need to process declarations before their uses.  `HandleMacros` must be done before `NameAnon` because macro usage may be used to determine how anonymous types are named.  `NameAnon` must come before `ResolveBindingSpec` because it enables users to configure anonymous types using our generated names for them.  `ResolveBindingSpec` must come before `HandleTypedefs` to enable users to configure if a specific `typedef` is squashed for not.  `MangleNames` must come last; it could be put into the `Hs` phase, but we have to draw the line somewhere.

Selection is done in the `Parse` pass.  Clang parses everything, but we only create a `Frontend` AST declaration if a declaration is selected.  Declarations that are not selected are therefore not provided (via `unitDecls`) to subsequent passes and phases.

Selection and external binding specifications work together to provide composability.  For example, consider a header `demo.h` that has `#include <stdint.h>` and makes use of type `int64_t`.

If everything is selected, then `unitDecls` includes the declarations from `stdint.h` (as well as transitive includes) and `demo.h`.  We generate a Haskell `newtype` for `int64_t`.  Multiple modules with generated `int64_t` types cannot work together (without coercion) because the types are distinct.

If only declarations in the main file are selected, but there is no external binding specification, then we would not generate a Haskell type for `int64_t` and any reference to that type would be a "type not found" error.

If only declarations in the main file are selected and the `base` external binding specification is used, then we get the desired result.  We would not generate a Haskell type for `int64_t`.  We would instead use an existing type from `base` (`Data.Int.Int64`).  Multiple modules generated like this all use the same type, providing composability.

Binding specification types are specified by the C "spelling" and a list of headers that may provide them.

A C "spelling" is just a C name that includes a `struct`, `union`, or `enum` prefix when necessary.  (There are cases where `struct foo` and `foo` need to be distinguished.)

A list of headers is required because of how some C libraries organize dependencies, a consequence of how macros work.  Taking an example from `libc`, the `int64_t` type is provided by `inttypes.h` as well as `stdint.h`.  The actual declaration is generally done in an "internal" header that is included by both of these.  An "internal" header is not part of the public API and may differ across implementations.  An external binding type match requires that the source location be transitively included by any of the listed headers.

When a type such as `int64_t` is not selected, it is not included in `unitDecls`, but we need to check for external bindings.  A `Type` reference to the type does not have sufficient information: it does not provide the source location, which is required in order to match external binding specification types.  This is why `NonSelectedDecls` is needed.  It keeps track of the source locations of non-selected declarations so that we can match external binding specification types.

**Comment by TravisCardwell** [comment]
**Selecting transitive dependencies**

Parent declarations are *not* processed before child declarations.  In general, the opposite is true: a type that is used by value must be declared before use.

Minimal example:

```c
#include <stdint.h>

struct foo {
  int64_t i;
};
```

We process the declarations from `stdint.h` (and transitive included headers) before we see the usage in `struct foo`.

I do not know of a way to select transitive dependencies within the `Parse` pass.  My understanding is that selection is currently done in the `Parse` pass for performance reasons: to avoid creating `Frontend` AST nodes for *everything*.  Perhaps this optimization is not possible if we support selection of transitive dependencies...  I shall give it some more thought.

**Comment by dschrempf** [comment]
Thank you for your detailed summary and explanation. I think I will add your summary of the passes as comments to the code.

> I do not know of a way to select transitive dependencies within the Parse pass. My understanding is that selection is currently done in the Parse pass for performance reasons: to avoid creating Frontend AST nodes for everything. Perhaps this optimization is not possible if we support selection of transitive dependencies... I shall give it some more thought.

Exactly, that is the same conclusion I also came to.

**Comment by edsko** [comment]
Yes, we already said that right? If we want to do program slicing in this way, we'll need the use-decl graph, so there we'll need to select as a separate pass. 

I'd be inclined to keep the existing selection mechanism in place as well though; I'm still worried that if we _always_ have to parse everything (because we are only selecting later), we might end up trying _and failing_ to parse stuff we don't actually need at all. Of course we'd have to think about how these two selection passes interact.

**Comment by dschrempf** [comment]
The basic implementation was done in https://github.com/well-typed/hs-bindgen/pull/775
We factored out two issues:
- Manual section: https://github.com/well-typed/hs-bindgen/issues/785
- Common selection predicate: https://github.com/well-typed/hs-bindgen/issues/784
